kratos/factories/base_factory.h

```diff
@@ -0,0 +1,417 @@
+    {
+        KRATOS_ERROR << "Calling the base class BaseFactory" << std::endl;
+    }
+
+    /**
+     * @brief This method is an auxiliar method to create a new class with Model and settings
+     * @param rModel The model containing the problem
+     * @param Settings The settings of the factory
+     * @return The pointer to the class of interest
+     */
+    virtual typename ClassType::Pointer CreateClass(Model& rModel, Kratos::Parameters Settings)  const
+    {
+        KRATOS_ERROR << "Calling the base class BaseFactory" << std::endl;
+    }
+
+    /**
+     * @brief This method is an auxiliar method to create a new class with model part and settings
+     * @param rModelPart The model part containing the problem
+     * @param Settings The settings of the factory
+     * @return The pointer to the class of interest
+     */
+    virtual typename ClassType::Pointer CreateClass(ModelPart& rModelPart, Kratos::Parameters Settings)  const
+    {
+        KRATOS_ERROR << "Calling the base class BaseFactory" << std::endl;
+    }
+
+    /**
+     * @brief This method is an auxiliar method to create a new class with auxiliar class and settings
+     * @param pAuxiliarClass The pointer to the auxiliar class
+     * @param Settings The settings of the factory
+     * @return The pointer to the class of interest
+     */
+    virtual typename ClassType::Pointer CreateClass(typename AuxiliarClassType::Pointer pAuxiliarClass, Kratos::Parameters Settings)  const
+    {
+        KRATOS_ERROR << "Calling the base class BaseFactory" << std::endl;
+    }
+};
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name Input and output
+///@{
+
+/// output stream function
+template<class TClass, typename TAuxiliarClass>
+inline std::ostream& operator << (std::ostream& rOStream,
+                                  const BaseFactory<TClass, TAuxiliarClass>& rThis)
+{
+    rThis.PrintInfo(rOStream);
+    rOStream << std::endl;
+    rThis.PrintData(rOStream);
+
+    return rOStream;
+}
+
+///@}
+
+typedef TUblasSparseSpace<double> SparseSpaceType;
+typedef TUblasDenseSpace<double> LocalSparseSpaceType;
+typedef LinearSolver<SparseSpaceType,LocalSparseSpaceType> LinearSolverType;
+
+typedef SolvingStrategy<SparseSpaceType, LocalSparseSpaceType, LinearSolverType> SolvingStrategyType;
+typedef BuilderAndSolver<SparseSpaceType, LocalSparseSpaceType, LinearSolverType> BuilderAndSolverType;
+typedef Scheme<SparseSpaceType,LocalSparseSpaceType> SchemeType;
+typedef ConvergenceCriteria<SparseSpaceType,LocalSparseSpaceType> ConvergenceCriteriaType;
+
+typedef BaseFactory<SolvingStrategyType> StrategyFactoryType;
+typedef BaseFactory<BuilderAndSolverType, LinearSolverType> BuilderAndSolverFactoryType;
+typedef BaseFactory<SchemeType> SchemeFactoryType;
+typedef BaseFactory<ConvergenceCriteriaType> ConvergenceCriteriaFactoryType;
+
+KRATOS_API_EXTERN template class KRATOS_API(KRATOS_CORE) KratosComponents<StrategyFactoryType>;
+
+#ifdef KRATOS_REGISTER_STRATEGY
```

Comment:
I am not autogenerating the name because @philbucher defended to consider the names as snake_case instead of CammelCase, the classes are CammelCase. I can do a function that transforms CammelCase to snake_case in order to do that

Reply:
I still defend the snake_case since this is our json style
also how would you get the class name fro the Object (this is the proposal right?)
=> afaik there is no portable way for getting the name of an object. There is `typeid(the_object).name` but this can contain mangling depending on the compiler

Reply:
Probably the easier way would be to define the method Name() so we can just call that.

Reply:
BTW @pooyan-dadvand technically what we register is the strategy, it its factory, but the strategy at the end

Reply:
> Meanwhile, you are not autogenerating the name, so I am not sure if being a macro is a good idea. Just consider that in the future, we will do the registration in each class using a dummy static variable like what is done in the tests.
> 
> I would consider the option of having a template method in the base class. (until we create the new registry class of the Kratos)

It is already templated

Reply:
wait a second! My comment was that we don't need the Macro! not taking out the name!!

In fact the name is necessary to hide the template parameters!

Reply:
I know that the base class is templated but I was saying that the create method should be templated